presto-main/src/main/java/com/facebook/presto/Session.java

```diff
@@ -99,6 +102,8 @@ public final class Session
     private final OptimizerInformationCollector optimizerInformationCollector = new OptimizerInformationCollector();
     private final OptimizerResultCollector optimizerResultCollector = new OptimizerResultCollector();
     private final CTEInformationCollector cteInformationCollector = new CTEInformationCollector();
+    private final Map<PlanNodeId, PlanNodeStatsEstimate> planNodeEstimateMap = new HashMap<>();
```

Comment:
Shouldn't this cache go to CachingCostProvider and CachingStatsProviderRespectfully?

Reply:
This will assume that plan node with the same node ID has the same estimation among different optimizers. I think this is a strong assumption which may not hold true. Currently we are constructing CachingStatsProvider in each optimizer and cache the stats within each optimizer itself.

Reply:
I think the session is a bit of a funny place for this (like it seems mostly that it's there because that's already getting passed around).  Maybe have some special stats info collector that we pass in to the optimizer and get the stats from.

Reply:
I tried to create an info collector and pass through optimizers. However I found it to be a huge change, as all optimizers, the CachingStatsProvider and CachingCostProvider, and code creating these instances need to change.
We already have a few information collectors in the Session class, I am wondering if we can keep these maps here to avoid the huge interface changes.